Microsoft.Azure.Devices.LongHaul.Module/IotHub.cs

```diff
@@ -0,0 +1,490 @@
+                        logger.Metric(DirectMethodToModuleRoundTripSeconds, sw.Elapsed.TotalSeconds);
+                        break;
+                    }
+                    catch (IotHubClientException ex) when (ex.ErrorCode == IotHubClientErrorCode.DeviceNotFound)
+                    {
+                        logger.Trace($"Caught exception invoking direct method.\n{ex}", TraceSeverity.Warning);
+                        // retry
+                    }
+                    catch (Exception ex)
+                    {
+                        logger.Trace($"Unexpected exception observed while invoking direct method.\n{ex}");
+                        break;
+                    }
+
+                    // retry delay
+                    await Task.Delay(s_retryInterval, ct).ConfigureAwait(false);
+                }
+
+                // interval delay
+                await Task.Delay(s_directMethodInvokeInterval, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Frequently send telemetry messages to the hub.
+        /// </summary>
+        /// <param name="ct">The cancellation token</param>
+        public async Task SendTelemetryMessagesAsync(Logger logger, CancellationToken ct)
+        {
+            // AMQP supports bulk telemetry sending, so we'll configure how many to send at a time.
+            int maxBulkMessages = _clientOptions.TransportSettings is IotHubClientAmqpSettings
+                ? 10
+                : 1;
+
+            // We want to test both paths for AMQP so we'll use a boolean to alternate between bulk and single.
+            bool sendSingle = true;
+
+            var pendingMessages = new List<TelemetryMessage>(maxBulkMessages);
+            bool loggedDisconnection = false;
+            logger.LoggerContext.Add(OperationName, LoggingConstants.TelemetryMessage);
+            Stopwatch sw = new();
+            while (!ct.IsCancellationRequested)
+            {
+                logger.Metric(ModuleMessageBacklog, _messagesToSend.Count);
+
+                // Wait when there are no messages to send, or if not connected
+                if (!IsConnected
+                    || !_messagesToSend.Any())
+                {
+                    try
+                    {
+                        await Task.Delay(s_messageLoopSleepTime, ct).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        // App is signalled to exit
+                        _logger.Trace($"Exit signal encountered. Terminating telemetry message pump.", TraceSeverity.Verbose);
+                        return;
+                    }
+                }
+
+                // If not connected, skip the work below this round.
+                if (!IsConnected)
+                {
+                    if (!loggedDisconnection)
+                    {
+                        loggedDisconnection = true;
```

Comment:
When does this get reset back to false when we've been reconnected?

Reply:
resetting it back to false when client is connected again.